[Product discount] Manual product discount flow

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->
## Manual product discount flow

This PR adds the possibility to apply a discount directly to an item in order when creating or editing an order.

Closes: #9352
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

We're using the order update endpoint and modifying `Order.Item.total` according to the discount amount provided by the user. More details about discounting using order update endpoint: p1655830808613139-slack-CGPNUU63E

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Create or edit an order containing the product(s) 
2. Click on an item and then "+ Add discount"
3. Enter a discount and return to the order edition screen
4. Verify order total is updated correctly and includes the discount amount

Focus on testing various scenarios including:
1. Multiple discounts 
2. Invalid discount amounts
3. Modifying the quantity of the discounted item
4. Try to find unhandled edge cases

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-android/assets/4527432/2b791791-6fe5-4aca-877c-9ba30b956220

### Roadmap
In the next PRs, we'll add:
1. Percentage discount picker
2. Showing discount amount values in the order create/edit screen on the items list
5. Improving the UI of the product details screen
6. Analytics events

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Implement a plain product discount flow

samiuelson committed:
Implement a plain product discount flow

samiuelson committed:
Update order details discount items text

samiuelson committed:
Request focus for discount textfield by default

samiuelson committed:
Update package declaration

samiuelson committed:
Add discount amount validation

samiuelson committed:
Fix total amount sync trigger in order creation

samiuelson committed:
Show discount amount in product details

samiuelson committed:
Hide "remove discount" button in case no discount added

samiuelson committed:
Improve discount amount processing

samiuelson committed:
Clean up code

samiuelson committed:
Fix detekt issues

samiuelson committed:
Add comment

samiuelson committed:
Introduce unit tests for discount edit view model

samiuelson committed:
Update unit tests

samiuelson committed:
Update unit tests

samiuelson committed:
Remove redundant code

samiuelson committed:
Fix UI tests

samiuelson committed:
Merge branch 'trunk' into 9352-product-discounts-full-flow

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/product/discount/GetItemDiscountAmountText.kt:
```diff
@@ -0,0 +1,13 @@
+    private val formatter: CurrencyFormatter
+) {
+    operator fun invoke(discountAmount: BigDecimal, currency: String): String {
+        return "-${formatter.formatCurrency(discountAmount, currency)}"
```
I think discounts usually shown without a minus sign, e.g.

Discount 5$, Discount 5% etc

samiuelson replied:
Fixed 0c3ef70

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/product/discount/OrderCreateEditProductDiscountScreen.kt:
```diff
@@ -0,0 +1,147 @@
+        {},
+        {},
+        {},
+        { _ -> }
```
np: isn't `{}` the same?

samiuelson replied:
Fixed ad7ad35

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/product/discount/OrderCreateEditProductDiscountScreen.kt:
```diff
@@ -0,0 +1,147 @@
+    onDiscountAmountChange: (String) -> Unit,
+) {
+    val state = viewState.collectAsState()
+    Scaffold(topBar = { Toolbar(onCloseClicked, onDoneClicked, state.value.isDoneButtonEnabled) }) { padding ->
```
Should it have elevation as other screens?

<img width="368" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/97af1e95-01fd-4c8c-80c4-aad03fa22850">

samiuelson replied:
Fixed in 7330471
<details>
<summary>Screenshot</summary>
<img src=https://github.com/woocommerce/woocommerce-android/assets/4527432/12df3954-a941-457d-b179-5a5ad705ee23 width=300/>
</details>

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/product/discount/OrderCreateEditProductDiscountViewModel.kt:
```diff
@@ -1,11 +1,106 @@
+    }
+
+    fun onDoneClicked() {
+        orderItem.updateAndGet {
```
That crashes if I click done when no discount entered

samiuelson replied:
Yeah, I fixed this crash in the next PR - https://github.com/woocommerce/woocommerce-android/pull/9395/files#diff-58d35608f7870da1a4edba728105a90b5b54a9d0b136a8852a80fd72d0c8b922R111

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/product/discount/OrderCreateEditProductDiscountScreen.kt:
```diff
@@ -0,0 +1,147 @@
+                    value = state.value.discountAmount,
+                    onValueChange = {
+                        if (it.isNotEmpty()) {
+                            onDiscountAmountChange(decimalTextProcessor.transformText(state.value.discountAmount, it))
```
We probably need to restrict entering any number in here, only "money" numbers, similar to what we have in IPP

https://github.com/woocommerce/woocommerce-android/assets/4923871/c5e70aea-5908-4210-b103-ba9f845df361

samiuelson replied:
Switched to "money"-style input field in 944942f

kidinov commented:
Not part of the PR and I could have mentioned that before, but just in case. Top and bottom margins here look different, and "edit" is not aligned too

<img width="403" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/6ab50d4c-92e6-48ca-a20c-41eb6ce3c9df">

samiuelson committed:
Clean up code

samiuelson committed:
Set TopAppBar's elevation to default value

kidinov commented:
Hm... not sure what's specific about this case (maybe taxes?) but the discount seems to be added instead of subtracted? Notice 83.6 price and 5$ discount -> total 88.6


https://github.com/woocommerce/woocommerce-android/assets/4923871/42db12bc-b289-4e1a-af7f-f2b41f2acb17


<img width="389" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/65e17439-731f-4fb9-8f73-a2030046f78d">


**UPD**
Oh, I got it. We show discounted prices already in the product details, not the original - that's confusing. Maybe to show the original one?

samiuelson committed:
Fix discount amount text

kidinov commented:
Seems to be it doesn't work well with the coupons. Does coupon override the discount? And discount actually doesn't work at all if there is a coupon used

Also, notice that we show the discount with `-` (and I think this is good in this case because it's a kind of column and it's easier to calculate) but we have a space between `-` and the amount on the order details screen and we don't have that on the editing screen

https://github.com/woocommerce/woocommerce-android/assets/4923871/0bea229b-5e8e-47bf-bbc0-959f80d12509

**UPD** 

Can confirm that on the web we use minus in the column and no sign when we indicate the discount amount:

<img width="466" alt="image" src="https://github.com/woocommerce/woocommerce-android/assets/4923871/6600b163-965f-4b28-bf78-fae4719c23a2">

kidinov reviewed (commented):
@samiuelson I left some comments and will take a look again tomorrow

samiuelson committed:
Improve discount input field

* Migrate to `WCOutlinedTypedTextField`
* Define `NullableCurrencyTextFieldValueMapper`

samiuelson commented:
Thanks for the review, @kidinov ! I've addressed your comments.

Let me share additional clarifications.

#### Discounts with coupons
This is a backend issue and it is on our radar. We decided to prevent adding discounts in case there are active coupons. More context here: p1689084213840619/1689083832.564069-slack-C025A8VV728.
This is still on a [To-Do list](https://github.com/woocommerce/woocommerce-android/issues/9352) and I'll tackle it in the next PR

#### Displaying discounts in Order create/edit screen
We decided to add an additional row to the Payments section displaying the total discount amount and show the item-specific discount amount in the items list below the item's total price. More context here: p1689084213840619/1689083832.564069-slack-C025A8VV728.
This is also on the [To-Do list](https://github.com/woocommerce/woocommerce-android/issues/9352) and will be added in the next PR

Let me know if this makes sense or if you see any other issues!

samiuelson committed:
Fix detekt issues

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/TypedTextField.kt:
```diff
@@ -177,6 +178,34 @@ class NullableBigDecimalTextFieldValueMapper(
     }
 }
 
+class NullableCurrencyTextFieldValueMapper(
```
Not for this PR, just to consider. We have similar logic in `WCMaterialOutlinedCurrencyEditTextView`, so maybe to make it reusable, because it seems like something we may want to have in many places in the app

kidinov reviewed (approved):
LGTM!

Do not forget to fix the UI stuff in the following PRs

samiuelson committed:
Merge remote-tracking branch 'origin/trunk' into 9352-product-discounts-full-flow

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/coupon/edit/OrderCreateCouponEditScreen.kt
